acceptance-tests/src/main/kotlin/org/ostelco/at/jersey/Tests.kt

```diff
@@ -1897,7 +1899,157 @@ class PlanTest {
+                    customerId = customerId,
+                    timeout = 30000L)
+
+            Thread.sleep(200)
+
+            purchaseRecords = get {
+                path = "/purchases"
+                this.email = email
+            }
+            assert(3 == purchaseRecords.size) { "Got ${purchaseRecords.size} purchase records, expected 3 after renewal" }
+        } finally {
+            StripePayment.deleteCustomer(customerId = customerId)
+        }
+    }
+
+    @Test
+    fun `jersey test - POST purchase plan with trial time and with insuffient funds payment source`() {
+
+        val email = "purchase-${randomInt()}@test.com"
+        val sku = "PLAN_10USD_DAY"
+        var customerId = ""
+
+        try {
+            customerId = createCustomer(name = "Test Purchase Plan User", email = email).id
+            enableRegion(email = email, region = "us")
+
+            val sourceId = StripePayment.createPaymentSourceId()
+            val insuffientFundsSourceId = StripePayment.createInsuffientFundsPaymentSourceId()
+            val newSourceId = StripePayment.createPaymentSourceId()
+
+            post<String> {
+                path = "/products/$sku/purchase"
+                this.email = email
+                queryParams = mapOf("sourceId" to sourceId)
+            }
+
+            // Wait for DB to be updated.
+            Thread.sleep(200)
+
+            var purchaseRecords: PurchaseRecordList = get {
+                path = "/purchases"
+                this.email = email
+            }
+            purchaseRecords.sortBy { it.timestamp }
+
+            assert(Instant.now().toEpochMilli() - purchaseRecords.last().timestamp < 10_000) { "Missing purchase record" }
+
+            // First record - created when trial time starts (cost is 0).
+            // Second record - actual charge record.
+            assert(2 == purchaseRecords.size) { "Got ${purchaseRecords.size} purchase records, expected 2" }
+            assertEquals(expectedPlanProductUS, purchaseRecords.last().product, "Incorrect 'Product' in purchase record")
+
+            // Switch to a "no funds" card.
+            // This should cause subscription renewal to fail.
+            post<PaymentSource> {
+                path = "/paymentSources"
+                this.email = email
+                queryParams = mapOf("sourceId" to insuffientFundsSourceId)
+            }
+            put<PaymentSource> {
+                path = "/paymentSources"
+                this.email = email
+                queryParams = mapOf("sourceId" to insuffientFundsSourceId)
+            }
+
+            // Actual charge for renewal will first be done after trial time
+            // expires, which will happen after 4 sec. Waiting a bit longer
+            // before checking the outcome.
+            StripeEventListener.waitForFailedSubscriptionRenewal(
+                    subscription = "stripe-event-jersey-purchase-fail-sub",
```

Comment:
This (along with other things) in this test is repeated code. Repeated code makes the code less readable in the sense that it's difficult to know what's special in something that looks similar but can't be known to be similar before a complete inspection has been done.  It is better to factor this out into a separate variable or method (depending on the context) so that it is obvious, at a glance, what is identical and what isn't, and also what the differences are (can be highlighted using parameters to methods)).